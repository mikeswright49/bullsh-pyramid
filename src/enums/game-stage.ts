export enum GameStage {
    Initiation,
    Memorization,
    Flipping,
    Declaration,
    Bullshit,
    Memory,
    Complete,
    Reveal,
}
