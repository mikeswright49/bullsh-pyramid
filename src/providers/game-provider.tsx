import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
