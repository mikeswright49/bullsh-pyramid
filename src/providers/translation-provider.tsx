import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
