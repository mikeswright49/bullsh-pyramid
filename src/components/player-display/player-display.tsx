import React, { useContext } from 'react';
import { GameStage } from 'src/enums/game-stage';
import { Hand } from 'src/components/hand/hand';
import { Player } from 'types/player';
import { GameContext } from 'src/context/game-context';
import { PlayerDeclaration } from './stages/player-declaration';
import { PlayerBullshit } from './stages/player-bullshit';
import { PlayerReveal } from './stages/player-reveal';

export function PlayerDisplay({ player }: { player: Player }) {
    const gameState = useContext(GameContext);
    const { gameStage } = gameState;

    if (!player || !gameState) {
        return null;
    }

    switch (gameStage) {
        case GameStage.Initiation:
            return (
                <>
                    <div className="stack-y-2">
                        <h3>Waiting for the game to start</h3>
                    </div>
                </>
            );
        case GameStage.Memorization:
            return (
                <>
                    <div className="stack-y-2">
                        <h3>You have 30s to remember your cards!</h3>
                        <div className="stack-y-2">
                            <Hand cards={player.hand} showSelector={false} />
                        </div>
                    </div>
                </>
            );
        case GameStage.Flipping:
            return (
                <>
                    <div className="stack-y-2">
                        <h3>Flipping a card</h3>
                        <div className="stack-y-2">
                            <Hand cards={player.hand} showSelector={false} />
                        </div>
                    </div>
                </>
            );
        case GameStage.Declaration:
            return <PlayerDeclaration player={player} />;
        case GameStage.Bullshit:
            return <PlayerBullshit player={player} />;
        case GameStage.Reveal:
            return <PlayerReveal />;
        case GameStage.Memory:
            return (
                <>
                    <div className="stack-y-2">
                        <h3>Well which cards did you have?</h3>
                    </div>
                </>
            );
        case GameStage.Complete:
            return (
                <>
                    <div className="stack-y-2">
                        <h3>Congrats you&apos;re soft in the head</h3>
                    </div>
                </>
            );
        default:
            return <h1>Terrible error</h1>;
    }
}
