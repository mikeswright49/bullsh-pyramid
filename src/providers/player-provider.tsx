import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
