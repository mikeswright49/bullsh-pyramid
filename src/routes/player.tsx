import React, { useContext } from 'react';
import { PlayerContext } from 'src/providers/player-provider';
import { HostDisplay } from 'src/components/host-display/host-display';
import { PlayerDisplay } from 'src/components/player-display/player-display';
import { GameBoard } from 'src/components/game-board/game-board';

export function PlayerRoute() {
    const player = useContext(PlayerContext);
    if (!player) {
        return null;
    }
    return (
        <div className="container" data-testid="player-route">
            <div className="row">
                <div className="col-3">
                    {player.isHost && <HostDisplay />}
                    <PlayerDisplay player={player} />
                </div>
                <div className="col-9">
                    <GameBoard />
                </div>
            </div>
        </div>
    );
}
