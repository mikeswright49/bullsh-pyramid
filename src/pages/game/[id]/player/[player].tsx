import React from 'react';
import { useRouter } from 'next/router';
import { Layout } from 'src/components/layout/layout';
import { GameProvider } from 'src/providers/game-provider';
import { PlayersProvider } from 'src/providers/players-provider';
import { PlayerProvider } from 'src/providers/player-provider';
import { PlayerRoute } from 'src/routes/player';
import { VotesProvider } from 'src/providers/votes-provider';
import { TranslationProvider } from 'src/providers/translation-provider';

export default function Player(): JSX.Element {
    const router = useRouter();
    const gameId: string = router.query.id as string;
    const playerId: string = router.query.player as string;

    if (!playerId) {
        return null;
    }

    return (
        <Layout>
            <GameProvider gameId={gameId}>
                <PlayersProvider gameId={gameId}>
                    <PlayerProvider playerId={playerId}>
                        <VotesProvider>
                            <TranslationProvider>
                                <PlayerRoute />
                            </TranslationProvider>
                        </VotesProvider>
                    </PlayerProvider>
                </PlayersProvider>
            </GameProvider>
        </Layout>
    );
}
