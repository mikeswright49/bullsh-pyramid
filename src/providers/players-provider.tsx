import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
