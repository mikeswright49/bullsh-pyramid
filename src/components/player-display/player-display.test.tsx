import React from 'react';
import { GameStage } from '../../enums/game-stage';
import { render, screen } from '@testing-library/react';
import { PlayerDisplay } from './player-display';
import { MOCK_PLAYER } from '../../__tests__/data/mock-player';
import { MOCK_GAME_STATE } from '../../__tests__/data/mock-game-state';
import { GameContext } from '../../providers/game-provider';
import { Player } from '../../../types/player';
import { GameState } from '../../../types/game-state';
import { PlayerContext } from '../../providers/player-provider';
describe('Player Display', () => {
    let playerMock: Player;
    let gameContextMock: GameState;
    beforeEach(() => {
        playerMock = Object.assign({}, MOCK_PLAYER);
        gameContextMock = Object.assign({}, MOCK_GAME_STATE);
    });

    const testCases = Object.keys(GameStage)
        .map((key) => (Number.isInteger(parseInt(key, 10)) ? null : [key]))
        .filter((x) => !!x);

    test.each(testCases)('it should be able to render the player view for %s', (stage) => {
        gameContextMock.gameStage = GameStage[stage];
        render(
            <GameContext.Provider value={gameContextMock}>
                <PlayerContext.Provider value={playerMock}>
                    <PlayerDisplay />
                </PlayerContext.Provider>
            </GameContext.Provider>
        );
        expect(screen.getByTestId('player-display')).toMatchSnapshot();
    });
});
