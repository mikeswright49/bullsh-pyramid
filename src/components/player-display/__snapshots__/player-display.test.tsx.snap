// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Display it should be able to render the player view for Assign 1`] = `
<div
  data-testid="player-display"
>
  <h3>
    You don't have this card, just sit back and relax
  </h3>
</div>
`;

exports[`Player Display it should be able to render the player view for Bullshit 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      No one says they had it
    </h3>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Complete 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      Game Over Everyone
    </h3>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Declaration 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      Quick 5s! Do you have this card?
    </h3>
    <div
      class="stack-y-2"
    >
      <h3>
        Your hand
      </h3>
      <div
        class="hand"
      >
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-7"
            style="background-position: -553px -246px;"
          />
          <button
            data-testid="7-2-select-button"
          >
            This one
          </button>
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-3-8"
            style="background-position: -632px -369px;"
          />
          <button
            data-testid="8-3-select-button"
          >
            This one
          </button>
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-1-11"
            style="background-position: -869px -123px;"
          />
          <button
            data-testid="11-1-select-button"
          >
            This one
          </button>
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-0"
            style="background-position: -0px -246px;"
          />
          <button
            data-testid="0-2-select-button"
          >
            This one
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Flipping 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      Flipping a card
    </h3>
    <div
      class="stack-y-2"
    >
      <h3>
        Your hand
      </h3>
      <div
        class="hand"
      >
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-7"
            style="background-position: -553px -246px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-3-8"
            style="background-position: -632px -369px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-1-11"
            style="background-position: -869px -123px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-0"
            style="background-position: -0px -246px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Initiation 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      Waiting for the game to start
    </h3>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Memorization 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      You have 30s to remember your cards!
    </h3>
    <div
      class="stack-y-2"
    >
      <h3>
        Your hand
      </h3>
      <div
        class="hand"
      >
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-7"
            style="background-position: -553px -246px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-3-8"
            style="background-position: -632px -369px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-1-11"
            style="background-position: -869px -123px;"
          />
        </div>
        <div>
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-0"
            style="background-position: -0px -246px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Memory 1`] = `
<div
  data-testid="player-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      Well which cards did you have?
    </h3>
  </div>
</div>
`;

exports[`Player Display it should be able to render the player view for Reveal 1`] = `
<div
  data-testid="player-display"
>
  <h4>
    No one called bullshit. Time to move along.
  </h4>
</div>
`;
